src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackDecoder.cs

```diff
@@ -214,7 +214,12 @@ public void Decode(ReadOnlySpan<byte> data, bool endHeaders, HeaderCallback onHe
 
                             if (_integerDecoder.StartDecode((byte)(b & ~DynamicTableSizeUpdateMask), DynamicTableSizeUpdatePrefix))
                             {
-                                // TODO: validate that it's less than what's defined via SETTINGS
+                                if (_integerDecoder.Value > _maxDynamicTableSize)
+                                {
+                                    // Dynamic table size update is too large.
+                                    throw new HPackDecodingException();
+                                }
```

Comment:
What happens if there's a race between the server sending too large a header index and it receiving our settings that dictate the max?

Reply:
If I'm reading the spec right, this should never happen:

> To avoid unnecessary latency, clients are permitted to send additional frames to the server immediately after sending the client connection preface, without waiting to receive the server connection preface.

I interpret this to mean *only* clients are allowed to send additional frames without waiting for a settings frame; a server should wait for the client's settings frame.

The only time we need to worry about this is, I think, if we ever support sending a new value for this in a subsequent settings frame; it's not clear exactly how this kind of race should be handled for any setting and seems like some leniency might be required.

Reply:
We should assume for now that we just won't support dynamic table entries. If/when we do, we will need to revisit a few things.

Reply:
(in other word, I'm agreeing with you @scalablecory)

Reply:
Not important for this change, but I found that the specific scenario of us allowing this maximum table size settings frame mid-connection is fairly well defined:

> This dynamic table size update MUST occur at the beginning of the first header block following the change to the dynamic table size.  In HTTP/2, this follows a settings acknowledgment (see Section 6.5.3 of [HTTP2]).

When the client sends a settings frame defining a new maximum, we do not immediately shrink our table or apply that limit.

Instead, the server is allowed to use or add new entries using the old maximum until they ACK the settings. After they ACK, we can shrink the table.

Next, we would enforce that the next header block received (if any) contains a dynamic table size update as its first instruction.